pepper-apis/src/main/java/org/broadinstitute/ddp/export/ParticipantsResultSetUtil.java

```diff
@@ -0,0 +1,128 @@
+        } finally {
+            if (resultset != null) {
+                resultset.close();
+            }
+        }
+    }
+
+    /**
+     * Extract data for all {@link Participant}-s of a study
+     */
+    public static List<Participant> extractParticipantsFromResultSet(
+            Handle handle,
+            StudyDto studyDto,
+            Stream<Participant> resultset,
+            Map<String, String> emailStore) {
+
+        Map<String, String> usersMissingEmails = new HashMap<>();
+
+        // extract Participants data
+        var instanceDao = handle.attach(ActivityInstanceDao.class);
+        Map<String, Participant> participants = resultset
+                .peek(pt -> extractParticipantData(pt, emailStore, usersMissingEmails, instanceDao))
+                .collect(Collectors.toMap(pt -> pt.getUser().getGuid(), pt -> pt));
+
+        // fetch Participants emails
+        if (!usersMissingEmails.isEmpty()) {
+            ExportUtil.fetchAndCacheAuth0Emails(handle, studyDto, usersMissingEmails.keySet(), emailStore)
+                    .forEach((auth0UserId, email) -> participants.get(usersMissingEmails.get(auth0UserId)).getUser().setEmail(email));
+        }
+
+        extractParticipantsNonDefaultAddresses(handle, participants);
+
+        return new ArrayList<>(participants.values());
+    }
+
+    /**
+     * Extract data for one {@link Participant}
+     */
+    private static void extractParticipantData(
+            Participant pt,
+            Map<String, String> emailStore,
+            Map<String, String> usersMissingEmails,
+            ActivityInstanceDao instanceDao) {
+
+        String auth0UserId = pt.getUser().getAuth0UserId();
+        if (StringUtils.isBlank(auth0UserId)) {
+            return;
+        }
+
+        String email = emailStore.get(auth0UserId);
+        if (email == null) {
+            usersMissingEmails.put(auth0UserId, pt.getUser().getGuid());
+        } else {
+            pt.getUser().setEmail(email);
+        }
+
+        Set<Long> instanceIds = pt.getAllResponses().stream()
+                .map(ActivityResponse::getId)
+                .collect(Collectors.toSet());
+
+        try (var stream = instanceDao.bulkFindSubstitutions(instanceIds)) {
+            stream.forEach(wrapper -> pt.putActivityInstanceSubstitutions(
+                    wrapper.getActivityInstanceId(), wrapper.unwrap()));
+        }
+    }
+
+    /**
+     * Fetch non-default (snapshotted) {@link MailAddress}'es for all given {@link Participant}'s
+     * and then reduce the fetched data to stream of participants with Map(Long,MailAddress) where key = instanceId.
+     * Save maps with MailAddresses to a corresponding Participants.
+     */
+    private static void extractParticipantsNonDefaultAddresses(Handle handle, Map<String, Participant> participants) {
+        Set<Long> participantIds = participants.values().stream().map(p -> p.getUser().getId()).collect(Collectors.toSet());
```

Comment:
For more efficiency, I recommend building this `Set` while iterating through the participants in `extractParticipantsFromResultSet()`. You can probably just add another `.peek()` to the stream pipeline. Even better, you can change `participants` to `Map<Long, Participant>` (id -> participant), so we can just do `participants.keySet()` and get all the participant ids for free.

Reply:
Fixed